src/BayesFilters/src/sigma_point.cpp

```diff
@@ -32,21 +53,263 @@ void bfl::sigma_point::unscented_weights(const unsigned int n, const double alph
 
-    MatrixXd A = svd.matrixU() * svd.singularValues().cwiseSqrt().asDiagonal();
+    for (std::size_t i = 0; i < state.components; i++)
+    {
+        JacobiSVD<MatrixXd> svd = state.covariance(i).jacobiSvd(ComputeThinU);
 
-    MatrixXd sigma_points(state.dim, (state.dim * 2) + 1);
+        MatrixXd A = svd.matrixU() * svd.singularValues().cwiseSqrt().asDiagonal();
 
-    sigma_points << VectorXd::Zero(state.dim), std::sqrt(c) * A, -std::sqrt(c) * A;
+        Ref<MatrixXd> sp = sigma_points.middleCols(((state.dim * 2) + 1) * i, ((state.dim * 2) + 1));
 
-    if (state.dim_linear > 0)
-        sigma_points.topRows(state.dim_linear).colwise() += state.mean().topRows(state.dim_linear);
+        sp << VectorXd::Zero(state.dim), std::sqrt(c) * A, -std::sqrt(c) * A;
 
-    if (state.dim_circular > 0)
-        sigma_points.bottomRows(state.dim_circular) = directional_add(sigma_points.bottomRows(state.dim_circular), state.mean().bottomRows(state.dim_circular));
+        if (state.dim_linear > 0)
+            sp.topRows(state.dim_linear).colwise() += state.mean(i).topRows(state.dim_linear);
+
+        if (state.dim_circular > 0)
+            sp.bottomRows(state.dim_circular) = directional_add(sigma_points.bottomRows(state.dim_circular), state.mean(i).bottomRows(state.dim_circular));
+    }
 
     return sigma_points;
 }
+
+
+std::tuple<bool, GaussianMixture, MatrixXd> bfl::sigma_point::unscented_transform
+(
+    const GaussianMixture& input,
+    const UTWeight& weight,
+    FunctionEvaluation function
+)
+{
+    /* Sample sigma points. */
+    MatrixXd input_sigma_points = sigma_point::sigma_point(input, weight.c);
+
+    /* Propagate sigma points */
+    Data fun_data;
+    bool valid_fun_data;
+    std::tie(valid_fun_data, fun_data) = function(input_sigma_points.cast<float>());
+
+    /* Stop here if function evaluation failed. */
+    if (!valid_fun_data)
+        return std::make_tuple(false, GaussianMixture(), MatrixXd(0, 0));
+
+    /* For now casting Data to MatrixXd. */
+    MatrixXd prop_sigma_points = (bfl::any::any_cast<MatrixXf&&>(std::move(fun_data))).cast<double>();
+
+    /* Initialize transformed gaussian. */
+    GaussianMixture output(input.components, prop_sigma_points.rows());
+
+    /* Initialize cross covariance matrix. */
+    MatrixXd cross_covariance(input.dim, output.dim * output.components);
+
+    /* Process all the components of the mixture. */
+    std::size_t base = ((input.dim * 2) + 1);
+    for (std::size_t i = 0; i < input.components; i++)
+    {
+        Ref<MatrixXd> input_sp = input_sigma_points.middleCols(base * i, base);
+        Ref<MatrixXd> prop_sp = prop_sigma_points.middleCols(base * i, base);
+
+        /* Evaluate the mean. */
+        output.mean(i).noalias() = prop_sp * weight.mean;
+
+        /* Evaluate the covariance. */
+        prop_sp.colwise() -= output.mean(i);
+        output.covariance(i).noalias() = prop_sp * weight.cov.asDiagonal() * prop_sp.transpose();
```

Comment:
Did you try using something like `(prop_sp.rowwise() * weight.cov) * prop_sp.transpose();` instead of expanding the weigths with `weight.cov.asDiagonal()`? Probably it is ok to do it with `2n + 1` sigma points, otherwise the diagonal expansion can be costly. However, I don't know whether there are better optimization around vector-matrix multiplications (very likely to happen when linking aginast BLAS) compared to element-wise vector multiplications or not. Do you know anything about it?

Reply:
Before using `asDiagonal` I had a search on `Eigen` website and found this:

>On arrays, we can also use the operators *=, /=, * and / to perform coefficient-wise multiplication and division column-wise or row-wise. These operators are not available on matrices because it is not clear what they would do. If you want multiply column 0 of a matrix mat with v(0), column 1 with v(1), and so on, then use mat = mat * v.asDiagonal().

The extract is taken from [here](https://eigen.tuxfamily.org/dox/group__TutorialReductionsVisitorsBroadcasting.html).

It seemed to me that it was the only option available then.

Reply:
However, the following could be an alternative
```
(prop_sp.array().colwise() * weight.covariance.array()).matrix() * prop_sp.transpose()
```

Reply:
I was referring to the second expression, but I reckon the first one (i.e. the one you implement) is much more readable. Let's keep it as it is.

Reply:
Ok, in some threads people suggest that `.asDiagonal` is an Eigen pseudo-expression, or a sort of _view_, and is such that when performing the matrix product, the expression is eventually converted to the one using Eigen arrays anyway.